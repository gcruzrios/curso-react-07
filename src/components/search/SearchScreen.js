import React, { useMemo } from 'react'
import queryString  from "query-string";
import { useLocation } from 'react-router-dom';
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q='' } = queryString.parse( location.search );

    

    const [ formValues, handleInputChange ] = useForm({ searchText: q});
    const { searchText } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])
 
    //const heroesFiltered = getHeroesByName(searchText);

    const handleSearch = (e) => {
        e.preventDefault();
        //console.log(searchText)
        history.push(`?q=${ searchText }`);
    }

   
    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5" key="uniqueId0">
                    <h4> Search Form</h4>
                    <hr />
                    <form onSubmit={ handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange = { handleInputChange }
                        />

                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...

                        </button>
                        
                            
                        
                    </form>    

                </div>

                <div className="col-7" key="uniqueId1">
                    <h4>
                        Results
                    </h4>
                    <hr />
                    {
                    (q ==='') &&
                       <div className="alert alert-info">
                           Search a hero
                       </div>         
                    }
                    { 
                    
                    (q!=='' && heroesFiltered.length === 0) && 
                    
                        <div className="alert alert-danger">

                            There is not a hero hero with { q }

                        </div>
                    }

                    { heroesFiltered.map ( hero => (
                        <HeroCard
                            hey ={ hero.id}
                            {...hero}
                        />
                    ))}
                </div>

            </div>
        </div>
    )
}
