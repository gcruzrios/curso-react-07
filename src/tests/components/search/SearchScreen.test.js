import { mount } from "enzyme"
import React from 'react';
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe('pruebas en SearchScreen', () => {
   
   
    test('debe de mostrarse correctamente con valores por defecto ', () => {
        
        const wrapper= mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is not a hero hero with');
    })
    

    test('debe de mostrar a Batman y el valor de query string ', () => {
        const wrapper= mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })
    

})
