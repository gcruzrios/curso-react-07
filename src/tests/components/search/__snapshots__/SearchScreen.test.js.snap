// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en SearchScreen debe de mostrar a Batman y el valor de query string  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card h-100 ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                   
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                     
                    Detective Comics #27
                     
                  </small>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    MÃ¡s...
                  </a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en SearchScreen debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is not a hero hero with 
      </div>
    </div>
  </div>
</div>
`;
